File: 141/HasCycle.go
package _141

import "github.com/raw34/leetcode/runtime"

func hasCycle(head *runtime.ListNode) bool {
	if head == nil {
		return false
	}

	cache := make(map[*runtime.ListNode]int, 0)

	for head != nil {
		if cache[head] > 1 {
			return true
		}
		cache[head]++
		head = head.Next
	}

	return false
}

File: solution/0141/HasCycle.php
<?php
class Solution {
    /**
     * @param ListNode $head
     * @return Boolean
     */
    function hasCycle($head) {
        if ($head === null) {
            return false;
        }

        $slow = $head;
        $fast = $head->next;

        while ($fast !== null) {
            if ($slow === $fast) {
                return true;
            }

            $slow = $slow->next;
            $fast = $fast->next;

            if ($fast !== null) {
                $fast = $fast->next;
            }
        }

        return false;
    }
}